In [39]:
import pandas as pd
import tensorflow as tf
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


In [10]:
df_churn= pd.read_csv('cust_churn.csv')# loading data

In [11]:
df_churn

ID_cu    Surname  Credit_score  Country     sex  Age  \
0     15634602   Hargrave           619   France  Female   42   
1     15647311       Hill           608    Spain  Female   41   
2     15619304       Onio           502   France  Female   42   
3     15701354       Boni           699   France  Female   39   
4     15737888   Mitchell           850    Spain  Female   43   
...        ...        ...           ...      ...     ...  ...   
9995  15606229   Obijiaku           771   France    Male   39   
9996  15569892  Johnstone           516   France    Male   35   
9997  15584532        Liu           709   France  Female   36   
9998  15682355  Sabbatini           772  Germany    Male   42   
9999  15628319     Walker           792   France  Female   28   

      Tenure_with_bank_year    Balance  NumOfProducts  OwnCard  \
0                         2       0.00              1        1   
1                         1   83807.86              1        0   
2                         8  159660.80              3        1   
3                         1       0.00              2        0   
4                         2  125510.82              1        1   
...                     ...        ...            ...      ...   
9995                      5       0.00              2        1   
9996                     10   57369.61              1        1   
9997                      7       0.00              1        0   
9998                      3   75075.31              2        1   
9999                      4  130142.79              1        1   

      IsActiveMember  estSalary  Exited  
0                  1  101348.88       1  
1                  1  112542.58       0  
2                  0  113931.57       1  
3                  0   93826.63       0  
4                  1   79084.10       0  
...              ...        ...     ...  
9995               0   96270.64       0  
9996               1  101699.77       0  
9997               1   42085.58       1  
9998               0   92888.52       1  
9999               0   38190.78       0  

[10000 rows x 13 columns]

In [12]:
df_churn.isna().sum()# checking for null data

ID_cu                    0
Surname                  0
Credit_score             0
Country                  0
sex                      0
Age                      0
Tenure_with_bank_year    0
Balance                  0
NumOfProducts            0
OwnCard                  0
IsActiveMember           0
estSalary                0
Exited                   0
dtype: int64

In [ ]:
# slicing df_churn into input X and output Y,  customer ID and Surname are not relevant as such in this case

In [13]:
X = df_churn.iloc[:, 2:12]


In [14]:
X

Credit_score  Country     sex  Age  Tenure_with_bank_year    Balance  \
0              619   France  Female   42                      2       0.00   
1              608    Spain  Female   41                      1   83807.86   
2              502   France  Female   42                      8  159660.80   
3              699   France  Female   39                      1       0.00   
4              850    Spain  Female   43                      2  125510.82   
...            ...      ...     ...  ...                    ...        ...   
9995           771   France    Male   39                      5       0.00   
9996           516   France    Male   35                     10   57369.61   
9997           709   France  Female   36                      7       0.00   
9998           772  Germany    Male   42                      3   75075.31   
9999           792   France  Female   28                      4  130142.79   

      NumOfProducts  OwnCard  IsActiveMember  estSalary  
0                 1        1               1  101348.88  
1                 1        0               1  112542.58  
2                 3        1               0  113931.57  
3                 2        0               0   93826.63  
4                 1        1               1   79084.10  
...             ...      ...             ...        ...  
9995              2        1               0   96270.64  
9996              1        1               1  101699.77  
9997              1        0               1   42085.58  
9998              2        1               0   92888.52  
9999              1        1               0   38190.78  

[10000 rows x 10 columns]

In [54]:
Y=df_churn[df_churn.columns[-1]]

In [55]:
Y

0       1
1       0
2       1
3       0
4       0
       ..
9995    0
9996    0
9997    1
9998    1
9999    0
Name: Exited, Length: 10000, dtype: int64

In [ ]:
# need to convert strings into numeric, there are 2 such fields sex and country

In [ ]:
X[]

In [17]:

Sex_num=pd.get_dummies(X['sex'],drop_first=True)

In [18]:
X=pd.concat([X,Sex_num],axis=1)

In [19]:
X=X.drop(['sex'],axis=1)

In [20]:
X

Credit_score  Country  Age  Tenure_with_bank_year    Balance  \
0              619   France   42                      2       0.00   
1              608    Spain   41                      1   83807.86   
2              502   France   42                      8  159660.80   
3              699   France   39                      1       0.00   
4              850    Spain   43                      2  125510.82   
...            ...      ...  ...                    ...        ...   
9995           771   France   39                      5       0.00   
9996           516   France   35                     10   57369.61   
9997           709   France   36                      7       0.00   
9998           772  Germany   42                      3   75075.31   
9999           792   France   28                      4  130142.79   

      NumOfProducts  OwnCard  IsActiveMember  estSalary  Male  
0                 1        1               1  101348.88     0  
1                 1        0               1  112542.58     0  
2                 3        1               0  113931.57     0  
3                 2        0               0   93826.63     0  
4                 1        1               1   79084.10     0  
...             ...      ...             ...        ...   ...  
9995              2        1               0   96270.64     1  
9996              1        1               1  101699.77     1  
9997              1        0               1   42085.58     0  
9998              2        1               0   92888.52     1  
9999              1        1               0   38190.78     0  

[10000 rows x 10 columns]

In [21]:
country_num=pd.get_dummies(X["Country"],drop_first=True)

In [22]:
X=pd.concat([X,country_num],axis=1)

In [23]:
X=X.drop(['Country'],axis=1)

In [24]:
X

Credit_score  Age  Tenure_with_bank_year    Balance  NumOfProducts  \
0              619   42                      2       0.00              1   
1              608   41                      1   83807.86              1   
2              502   42                      8  159660.80              3   
3              699   39                      1       0.00              2   
4              850   43                      2  125510.82              1   
...            ...  ...                    ...        ...            ...   
9995           771   39                      5       0.00              2   
9996           516   35                     10   57369.61              1   
9997           709   36                      7       0.00              1   
9998           772   42                      3   75075.31              2   
9999           792   28                      4  130142.79              1   

      OwnCard  IsActiveMember  estSalary  Male  Germany  Spain  
0           1               1  101348.88     0        0      0  
1           0               1  112542.58     0        0      1  
2           1               0  113931.57     0        0      0  
3           0               0   93826.63     0        0      0  
4           1               1   79084.10     0        0      1  
...       ...             ...        ...   ...      ...    ...  
9995        1               0   96270.64     1        0      0  
9996        1               1  101699.77     1        0      0  
9997        0               1   42085.58     0        0      0  
9998        1               0   92888.52     1        1      0  
9999        1               0   38190.78     0        0      0  

[10000 rows x 11 columns]

In [28]:
X['Age'].value_counts()

37    478
38    477
35    474
36    456
34    447
     ... 
92      2
82      1
88      1
85      1
83      1
Name: Age, Length: 70, dtype: int64

In [47]:
Y.value_counts()  # 2 values so logistic regression can also be used

0    7963
1    2037
Name: Exited, dtype: int64

In [30]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.3, random_state = 0)
# train test split with 30%



In [31]:
X_train

Credit_score  Age  Tenure_with_bank_year    Balance  NumOfProducts  \
7681           641   33                      2  146193.60              2   
9031           541   39                      7       0.00              2   
3691           590   76                      5  160979.68              1   
202            516   50                      5       0.00              1   
5625           508   60                      7  143262.04              1   
...            ...  ...                    ...        ...            ...   
9225           594   32                      4  120074.97              2   
4859           794   22                      4  114440.24              1   
3264           738   35                      5  161274.05              2   
9845           590   38                      9       0.00              2   
2732           623   48                      1  108076.33              1   

      OwnCard  IsActiveMember  estSalary  Male  Germany  Spain  
7681        1               1   55796.83     1        0      0  
9031        1               0   19823.02     1        0      0  
3691        0               1   13848.58     0        0      0  
202         0               1  146145.93     1        0      1  
5625        1               1  129562.74     0        0      0  
...       ...             ...        ...   ...      ...    ...  
9225        1               1  162961.79     0        1      0  
4859        1               1  107753.07     0        0      1  
3264        1               0  181429.87     1        0      0  
9845        1               1  148750.16     0        0      1  
2732        1               0  118855.26     0        1      0  

[7000 rows x 11 columns]

In [34]:
X['estSalary'].max()


199992.48

In [35]:
X['estSalary'].min()

11.58

In [36]:
X['Balance'].max()

250898.09

In [37]:
X['Balance'].min()

0.0

In [40]:
#range of salary and Balance is large , so need to do scaling
scaled_value = StandardScaler()
X_train = scaled_value.fit_transform(X_train)
X_test = scaled_value.transform(X_test)

# ----Model 1 - logistic regression--

In [128]:
# model 1 using logistic regression

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score
import matplotlib.pyplot as plt

In [169]:
model1= LogisticRegression()

In [170]:
model1.fit(X_train,y_train)# fitting X train and yTrain to Logistic regression

LogisticRegression()

In [171]:
y_pred_log=model1.predict(X_test)

In [173]:
Cm_logreg= confusion_matrix(y_test, y_pred_log)
Cm_logreg

array([[2278,  101],
       [ 479,  142]])

In [168]:
model1.score(X_test,y_test)

0.8066666666666666

Score of logistic regression is 80.66%

# ---- Model 2-   Neuron network ----

In [134]:

from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential



In [136]:
from tensorflow.keras.layers import LeakyReLU,PReLU

In [137]:
from tensorflow.keras.layers import Dropout

In [183]:
Model2 = Sequential()

In [184]:
# there. are 11 input , next layer considering 8 nodes and output 1 node

Model2.add(Dense(units=11,activation='relu'))

In [185]:
Model2.add(Dense(units=8,activation='relu'))

In [186]:
Model2.add(Dense(units=1,activation='relu'))

In [187]:
Model2.compile(
              loss='binary_crossentropy',optimizer='adam',
              metrics=['accuracy'])

In [188]:
internal_model=Model2.fit(X_train,y_train,validation_split=0.25,batch_size=10,epochs=10)# with less epoch and batch 10

Epoch 1/10
525/525 [==============================] - 1s 2ms/step - loss: 1.2072 - accuracy: 0.7589 - val_loss: 0.6619 - val_accuracy: 0.7686
Epoch 2/10
525/525 [==============================] - 1s 2ms/step - loss: 0.6256 - accuracy: 0.7882 - val_loss: 0.5428 - val_accuracy: 0.8051
Epoch 3/10
525/525 [==============================] - 1s 2ms/step - loss: 0.5278 - accuracy: 0.8036 - val_loss: 0.5284 - val_accuracy: 0.8194
Epoch 4/10
525/525 [==============================] - 1s 2ms/step - loss: 0.4950 - accuracy: 0.8164 - val_loss: 0.4911 - val_accuracy: 0.8206
Epoch 5/10
525/525 [==============================] - 1s 2ms/step - loss: 0.4603 - accuracy: 0.8244 - val_loss: 0.4750 - val_accuracy: 0.8200
Epoch 6/10
525/525 [==============================] - 1s 2ms/step - loss: 0.4314 - accuracy: 0.8293 - val_loss: 0.4645 - val_accuracy: 0.8223
Epoch 7/10
525/525 [==============================] - 1s 2ms/step - loss: 0.4246 - accuracy: 0.8265 - val_loss: 0.4455 - val_accuracy: 0.8240
Epoch 

In [189]:
internal_model=Model2.fit(X_train,y_train,validation_split=0.25,batch_size=10,epochs=20)
# with 20 Epoch accuracy increased

Epoch 1/20
525/525 [==============================] - 1s 2ms/step - loss: 0.3904 - accuracy: 0.8392 - val_loss: 0.4078 - val_accuracy: 0.8360
Epoch 2/20
525/525 [==============================] - 1s 2ms/step - loss: 0.3796 - accuracy: 0.8415 - val_loss: 0.4096 - val_accuracy: 0.8389
Epoch 3/20
525/525 [==============================] - 1s 2ms/step - loss: 0.3772 - accuracy: 0.8465 - val_loss: 0.4318 - val_accuracy: 0.8446
Epoch 4/20
525/525 [==============================] - 1s 2ms/step - loss: 0.3791 - accuracy: 0.8482 - val_loss: 0.4259 - val_accuracy: 0.8389
Epoch 5/20
525/525 [==============================] - 1s 2ms/step - loss: 0.3766 - accuracy: 0.8488 - val_loss: 0.4231 - val_accuracy: 0.8406
Epoch 6/20
525/525 [==============================] - 1s 2ms/step - loss: 0.3751 - accuracy: 0.8467 - val_loss: 0.4359 - val_accuracy: 0.8406
Epoch 7/20
525/525 [==============================] - 1s 2ms/step - loss: 0.3924 - accuracy: 0.8381 - val_loss: 0.4202 - val_accuracy: 0.8354
Epoch 

In [190]:
internal_model=Model2.fit(X_train,y_train,validation_split=0.25,batch_size=10,epochs=50)
# with 50 Epoch accuracy increased abit not much to 85

Epoch 1/50
525/525 [==============================] - 1s 2ms/step - loss: 0.3517 - accuracy: 0.8549 - val_loss: 0.4183 - val_accuracy: 0.8474
Epoch 2/50
525/525 [==============================] - 1s 2ms/step - loss: 0.3497 - accuracy: 0.8520 - val_loss: 0.4305 - val_accuracy: 0.8406
Epoch 3/50
525/525 [==============================] - 1s 2ms/step - loss: 0.3558 - accuracy: 0.8522 - val_loss: 0.4136 - val_accuracy: 0.8491
Epoch 4/50
525/525 [==============================] - 1s 2ms/step - loss: 0.3507 - accuracy: 0.8549 - val_loss: 0.4365 - val_accuracy: 0.8451
Epoch 5/50
525/525 [==============================] - 1s 2ms/step - loss: 0.3527 - accuracy: 0.8554 - val_loss: 0.4143 - val_accuracy: 0.8423
Epoch 6/50
525/525 [==============================] - 1s 2ms/step - loss: 0.3453 - accuracy: 0.8545 - val_loss: 0.4335 - val_accuracy: 0.8509
Epoch 7/50
525/525 [==============================] - 1s 2ms/step - loss: 0.3481 - accuracy: 0.8543 - val_loss: 0.4154 - val_accuracy: 0.8509
Epoch 

In [191]:
y_pred = Model2.predict(X_test)

94/94 [==============================] - 0s 1ms/step


In [192]:
y_pred

array([[0.45319575],
       [0.35199225],
       [0.17543064],
       ...,
       [0.08757344],
       [0.34100422],
       [0.52544314]], dtype=float32)

In [197]:
y_pred_NN = (y_pred > 0.5)
y_pred_NN

array([[False],
       [False],
       [False],
       ...,
       [False],
       [False],
       [ True]])

In [198]:
Cm_NN= confusion_matrix(y_test, y_pred_NN)
Cm_NN

array([[2220,  159],
       [ 273,  348]])

In [199]:
score=accuracy_score(y_pred_NN,y_test)
score

0.856

In [200]:
score1=accuracy_score(y_test,y_pred_NN)
score1

0.856

Neural Netwrok performed better than logistic regression, NN score is 85.6%

# --- Model 3 Decision Tree and ensemble--

In [152]:
from sklearn import tree, ensemble

In [201]:
Model3= tree.DecisionTreeClassifier()
Model3.fit(X_train,y_train)

DecisionTreeClassifier()

In [202]:
y_pred_DT = Model3.predict(X_test)

In [203]:
y_pred_DT

array([0, 1, 0, ..., 0, 0, 0])

In [204]:
Cm_DT= confusion_matrix(y_test, y_pred_DT)
Cm_DT

array([[2051,  328],
       [ 286,  335]])

In [205]:
Model3.score(X_test, y_test)

0.7953333333333333

In [206]:
score=accuracy_score(y_pred_DT,y_test)
score

0.7953333333333333

###### Decision Tree has the accuracy of 80.16 and lowest with default setting, but as decide the depth 2 and increase depth till 5 acuracy increasing  till 85.9, higher depth is overfitting scenarion and not genralised

In [207]:
Model3_2= tree.DecisionTreeClassifier(max_depth=2)
Model3_2.fit(X_train,y_train)

DecisionTreeClassifier(max_depth=2)

In [208]:
Model3_2.score(X_test, y_test)

0.836

In [209]:
Model3_3= tree.DecisionTreeClassifier(max_depth=3)
Model3_3.fit(X_train,y_train)
Model3_3.score(X_test, y_test)

0.84

In [210]:
Model3_4= tree.DecisionTreeClassifier(max_depth=4)
Model3_4.fit(X_train,y_train)
Model3_4.score(X_test, y_test)

0.8546666666666667

In [211]:
Model3_5= tree.DecisionTreeClassifier(max_depth=5)
Model3_5.fit(X_train,y_train)
Model3_5.score(X_test, y_test)

0.8586666666666667

# So Neural network model 2 performed better 